design/src/client/orchestrator.md

```diff
@@ -0,0 +1,205 @@
+
+## How is an orchestrator configured?
+
+While the structure of an orchestrator is fixed, the actions it takes during its lifecycle are highly configurable. Users have two ways to configure this process:
+
+- **Runtime Plugins**:
+    - **When can these be set?** Any time before calling `orchestrate`.
+    - **When are they called by the orchestrator?** In two batches, at the very beginning of `orchestrate`.
+    - **What can they do?**
+        - They can set configuration to be used by the orchestrator or in interceptors.
+        - They can set interceptors.
+    - **Are they user-definable?** No. At present, only smithy maintainers may define these.
+- **Interceptors**:
+    - **When can these be set?** Any time before calling `orchestrate`.
+    - **When are they called by the orchestrator?** At each step in the request-response lifecycle.
+    - **What can they do?**
+        - They can set configuration to be used by the orchestrator or in interceptors.
+        - They can log information.
+        - Depending on when they're run, they can modify the input, transmit request, transmit response, and the output/error.
+    - **Are they user-definable?** Yes.
+
+Configuration for a request is constructed by runtime plugins just after calling `orchestrate`. Configuration is stored in a `ConfigBag`: a hash map that's keyed on type's [`TypeId`][TypeId] (an opaque object, managed by the Rust compiler, which references some type.)
+
+## What does the orchestrator do?
+
+The orchestrator's work is divided into four phases:
+
+*NOTE: If an interceptor fails, then the other interceptors for that lifecycle event are still run. All resulting errors are collected and emitted together.*
+
+0. **Building the `ConfigBag` and mounting interceptors**.
+    - *This phase is infallible.*
+    - An interceptor context is created. This will hold request and response objects, making them available to interceptors.
+    - All runtime plugins set at the client-level are run. These plugins can set config and mount interceptors. Any _"read before execution"_ interceptors that have been set get run.
+    - All runtime plugins set at the operation-level are run. These plugins can also set config and mount interceptors. Any new _"read before execution"_ interceptors that have been set get run.
+1. **Request Construction**
+    - *This phase is fallible.*
+    - The _"read before serialization"_ and _"modify before serialization"_ interceptors are called.
+    - The input is serialized into a transmit request.
+    -  The _"read after serialization"_ and _"modify before retry_ loop" interceptors are called.
+    - Before making an attempt, the retry handler is called to check if an attempt should be made. The retry handler makes this decision for an initial attempt as well as for the retry attempts. If an initial attempt should be made, then the orchestrator enters the Dispatch phase. Otherwise, a throttling error is returned.
```

Comment:
Seems like there can be wasted effort if we decide this after serialization.

Reply:
The internal spec defines it this way. I'll follow up and see if that's deliberate or just an oversight.